File: packages/typescript/tests/validators.test.ts
import {
  luhnCheck,
  validateEmail,
  validatePaymentIntent,
  validateByCountry,
  validateCardNumber,
  validateExpirationDate,
} from '../src/utils/validators';
import { ValidationException, InvalidCardException } from '../src/errors/index';
import type { PaymentIntent } from '../src/types/index';

const baseIntent: PaymentIntent = {
  amount: 100,
  currency: 'COP',
  customer: {
    fullName: 'Juan Pérez',
    email: 'juan@example.com',
    documentType: 'CC',
    documentNumber: '12345678',
  },
};

describe('luhnCheck', () => {
  it('validates a correct Visa test number', () => {
    expect(luhnCheck('4111111111111111')).toBe(true);
  });

  it('validates a correct Mastercard test number', () => {
    expect(luhnCheck('5500005555555559')).toBe(true);
  });

  it('rejects an invalid card number', () => {
    expect(luhnCheck('1234567890123456')).toBe(false);
  });

  it('rejects a number with wrong check digit', () => {
    expect(luhnCheck('4111111111111112')).toBe(false);
  });

  it('rejects too-short numbers', () => {
    expect(luhnCheck('41111111')).toBe(false);
  });

  it('handles numbers with spaces', () => {
    // 4111 1111 1111 1111 - Luhn valid
    expect(luhnCheck('4111 1111 1111 1111')).toBe(true);
  });
});

describe('validateEmail', () => {
  it('validates a correct email', () => {
    expect(validateEmail('user@example.com')).toBe(true);
  });

  it('validates email with subdomain', () => {
    expect(validateEmail('user@mail.example.co')).toBe(true);
  });

  it('rejects email without @', () => {
    expect(validateEmail('notanemail')).toBe(false);
  });

  it('rejects email without domain', () => {
    expect(validateEmail('user@')).toBe(false);
  });

  it('rejects email with spaces', () => {
    expect(validateEmail('user @example.com')).toBe(false);
  });
});

describe('validatePaymentIntent', () => {
  it('passes for a valid intent', () => {
    expect(() => validatePaymentIntent(baseIntent, 12345)).not.toThrow();
  });

  it('throws for amount = 0', () => {
    expect(() => validatePaymentIntent({ ...baseIntent, amount: 0 }, 12345)).toThrow(ValidationException);
  });

  it('throws for negative amount', () => {
    expect(() => validatePaymentIntent({ ...baseIntent, amount: -10 }, 12345)).toThrow(ValidationException);
  });

  it('throws for unsupported currency', () => {
    expect(() =>
      validatePaymentIntent({ ...baseIntent, currency: 'XYZ' }, 12345),
    ).toThrow(ValidationException);
  });

  it('throws when customer is missing', () => {
    const intent = { ...baseIntent, customer: undefined as unknown as typeof baseIntent.customer };
    expect(() => validatePaymentIntent(intent, 12345)).toThrow(ValidationException);
  });

  it('throws for invalid email', () => {
    const intent = {
      ...baseIntent,
      customer: { ...baseIntent.customer, email: 'not-an-email' },
    };
    expect(() => validatePaymentIntent(intent, 12345)).toThrow(ValidationException);
  });

  it('throws when etyCode is missing', () => {
    expect(() => validatePaymentIntent(baseIntent, undefined)).toThrow(ValidationException);
  });

  it('accepts USD currency', () => {
    expect(() =>
      validatePaymentIntent({ ...baseIntent, currency: 'USD' }, 12345),
    ).not.toThrow();
  });
});

describe('validateByCountry', () => {
  it('passes valid CO intent', () => {
    expect(() =>
      validateByCountry(
        {
          ...baseIntent,
          customer: { ...baseIntent.customer, documentType: 'CC' },
          paymentSystem: '1',
        },
        'CO',
      ),
    ).not.toThrow();
  });

  it('throws for invalid document type in CO', () => {
    expect(() =>
      validateByCountry(
        {
          ...baseIntent,
          customer: { ...baseIntent.customer, documentType: 'CURP' },
          paymentSystem: '1',
        },
        'CO',
      ),
    ).toThrow(ValidationException);
  });

  it('throws for PSE without UserType in CO', () => {
    expect(() =>
      validateByCountry(
        {
          ...baseIntent,
          paymentSystem: '0',
          // userType missing
        },
        'CO',
      ),
    ).toThrow(ValidationException);
  });

  it('passes for PSE with UserType in CO', () => {
    expect(() =>
      validateByCountry(
        {
          ...baseIntent,
          paymentSystem: '0',
          userType: '0',
        },
        'CO',
      ),
    ).not.toThrow();
  });

  it('throws for SPEI with token in MX', () => {
    expect(() =>
      validateByCountry(
        {
          ...baseIntent,
          paymentSystem: '7',
          tokenId: 'tok_123',
          customer: { ...baseIntent.customer, documentType: 'RFC' },
        },
        'MX',
      ),
    ).toThrow(ValidationException);
  });

  it('passes for unknown country (no validation)', () => {
    expect(() =>
      validateByCountry({ ...baseIntent }, 'AR'),
    ).not.toThrow();
  });
});

describe('validateCardNumber', () => {
  it('passes for valid card', () => {
    expect(() => validateCardNumber('4111111111111111')).not.toThrow();
  });

  it('throws for invalid card', () => {
    expect(() => validateCardNumber('1234567890123456')).toThrow(InvalidCardException);
  });
});

describe('validateExpirationDate', () => {
  const nextYear = new Date().getFullYear() + 1;

  it('passes for future date', () => {
    expect(() => validateExpirationDate(`12/${nextYear}`)).not.toThrow();
  });

  it('throws for past date', () => {
    expect(() => validateExpirationDate('01/2020')).toThrow(InvalidCardException);
  });

  it('throws for invalid format', () => {
    expect(() => validateExpirationDate('2025-12')).toThrow(InvalidCardException);
  });

  it('throws for invalid month', () => {
    expect(() => validateExpirationDate(`13/${nextYear}`)).toThrow(InvalidCardException);
  });
});

File: packages/kotlin/src/test/kotlin/com/ecollect/sdk/ValidatorsTest.kt
package com.ecollect.sdk

import com.ecollect.sdk.exceptions.InvalidCardException
import com.ecollect.sdk.exceptions.ValidationException
import com.ecollect.sdk.types.PaymentIntent
import com.ecollect.sdk.utils.Validators
import org.junit.Assert.*
import org.junit.Test

class ValidatorsTest {

    // --- Luhn check tests ---

    @Test
    fun `luhnCheck returns true for valid VISA test card`() {
        assertTrue(Validators.luhnCheck("4111111111111111"))
    }

    @Test
    fun `luhnCheck returns true for valid Mastercard test card`() {
        assertTrue(Validators.luhnCheck("5500005555555559"))
    }

    @Test
    fun `luhnCheck returns true for valid AMEX test card`() {
        assertTrue(Validators.luhnCheck("371449635398431"))
    }

    @Test
    fun `luhnCheck returns false for invalid card number`() {
        assertFalse(Validators.luhnCheck("1234567890123456"))
    }

    @Test
    fun `luhnCheck returns false for all zeros`() {
        assertFalse(Validators.luhnCheck("0000000000000000"))
    }

    @Test
    fun `luhnCheck returns false for too short number`() {
        assertFalse(Validators.luhnCheck("12345"))
    }

    @Test
    fun `luhnCheck strips spaces and dashes`() {
        // "4111 1111 1111 1111" -> "4111111111111111"
        assertTrue(Validators.luhnCheck("4111111111111111"))
    }

    // --- validateCardNumber tests ---

    @Test(expected = InvalidCardException::class)
    fun `validateCardNumber throws for invalid card`() {
        Validators.validateCardNumber("1234567890123456")
    }

    @Test
    fun `validateCardNumber passes for valid card`() {
        // Should not throw
        Validators.validateCardNumber("4111111111111111")
    }

    // --- validateExpirationDate tests ---

    @Test
    fun `validateExpirationDate passes for future date`() {
        Validators.validateExpirationDate("12/2030")
    }

    @Test(expected = InvalidCardException::class)
    fun `validateExpirationDate throws for past date`() {
        Validators.validateExpirationDate("01/2020")
    }

    @Test(expected = InvalidCardException::class)
    fun `validateExpirationDate throws for wrong format`() {
        Validators.validateExpirationDate("2030/12")
    }

    @Test(expected = InvalidCardException::class)
    fun `validateExpirationDate throws for invalid month 13`() {
        Validators.validateExpirationDate("13/2030")
    }

    // --- validatePaymentIntent tests ---

    @Test(expected = ValidationException::class)
    fun `validatePaymentIntent throws for zero amount`() {
        val intent = PaymentIntent(
            amount = 0.0,
            currency = "COP",
            referenceArray = listOf("ref1")
        )
        Validators.validatePaymentIntent(intent)
    }

    @Test(expected = ValidationException::class)
    fun `validatePaymentIntent throws for negative amount`() {
        val intent = PaymentIntent(
            amount = -50.0,
            currency = "COP",
            referenceArray = listOf("ref1")
        )
        Validators.validatePaymentIntent(intent)
    }

    @Test(expected = ValidationException::class)
    fun `validatePaymentIntent throws for blank currency`() {
        val intent = PaymentIntent(
            amount = 100.0,
            currency = "",
            referenceArray = listOf("ref1")
        )
        Validators.validatePaymentIntent(intent)
    }

    @Test(expected = ValidationException::class)
    fun `validatePaymentIntent throws for empty referenceArray`() {
        val intent = PaymentIntent(
            amount = 100.0,
            currency = "COP",
            referenceArray = emptyList()
        )
        Validators.validatePaymentIntent(intent)
    }

    @Test
    fun `validatePaymentIntent passes for valid intent`() {
        val intent = PaymentIntent(
            amount = 50000.0,
            currency = "COP",
            referenceArray = listOf("CC", "12345678", "ORD-001", "John Doe", "j@test.com", "3001234567")
        )
        // Should not throw
        Validators.validatePaymentIntent(intent)
    }

    // --- validateByCountry tests ---

    @Test
    fun `validateByCountry passes for CC in Colombia`() {
        Validators.validateByCountry("CC", "CO")
    }

    @Test
    fun `validateByCountry passes for CI in Dominican Republic`() {
        Validators.validateByCountry("CI", "DO")
    }

    @Test
    fun `validateByCountry passes for RFC in Mexico`() {
        Validators.validateByCountry("RFC", "MX")
    }

    @Test(expected = ValidationException::class)
    fun `validateByCountry throws for CC in Mexico`() {
        Validators.validateByCountry("CC", "MX")
    }

    @Test(expected = ValidationException::class)
    fun `validateByCountry throws for unsupported country`() {
        Validators.validateByCountry("ID", "US")
    }

    @Test
    fun `validateByCountry passes for PP in all countries`() {
        Validators.validateByCountry("PP", "CO")
        Validators.validateByCountry("PP", "DO")
        Validators.validateByCountry("PP", "MX")
    }
}
